File: 026_MiddleOfLinkedList.cpp
/*
Problem: Find Middle of Linked List
Description: Find the middle node of a linked list. If there are two middle nodes,
             return the second middle node.

Approach: Use slow and fast pointer (tortoise and hare).
          When fast reaches end, slow is at middle.
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        return slow;
    }
};

int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    
    Solution sol;
    ListNode* middle = sol.middleNode(head);
    
    cout << "Middle Node: " << middle->val << endl;
    
    return 0;
}

/*
Time Complexity: O(n) where n is the number of nodes
Space Complexity: O(1)
*/

File: 026_MiddleOfLinkedList.py
"""
Problem: Find Middle of Linked List
Description: Find the middle node of a linked list. If there are two middle nodes,
             return the second middle node.

Approach: Use slow and fast pointer (tortoise and hare).
          When fast reaches end, slow is at middle.
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        return slow

if __name__ == "__main__":
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    
    sol = Solution()
    middle = sol.middleNode(head)
    
    print(f"Middle Node: {middle.val}")

"""
Time Complexity: O(n) where n is the number of nodes
Space Complexity: O(1)
"""
